core/src/main/scala/org/apache/spark/util/SizeEstimator.scala

```diff
@@ -89,7 +90,13 @@ object SizeEstimator extends Logging {
 
   // A cache of ClassInfo objects for each class
   // We use weakKeys to allow GC of dynamically created classes
-  private val classInfos = new MapMaker().weakKeys().makeMap[Class[_], ClassInfo]()
+  private val classInfos = new ThreadLocal[WeakHashMap[Class[_], ClassInfo]] {
+    override def initialValue(): java.util.WeakHashMap[Class[_], ClassInfo] = {
+      val toReturn = new WeakHashMap[Class[_], ClassInfo]()
+      toReturn.put(classOf[Object], new ClassInfo(objectSize, new Array[Int](0)))
+      return toReturn
```

Comment:
What I meant was, continue to use ThreadLocal, but maintain the MapMaker's result for thlocal.get()

And move the initilization to initialize() instead of in initialValue()